crates/aggchain-proof-core/src/bridge/mod.rs

```diff
@@ -394,27 +399,38 @@ impl BridgeConstraintsInput {
         self.verify_constrained_global_indices()?;
         self.verify_inserted_gers()
     }
-}
 
-/// Validate that the rebuilt hash chain is equal to the new hash chain.
-fn validate_hash_chain(
-    hashes: impl Iterator<Item = Digest>,
-    prev_hash_chain: Digest,
-    new_hash_chain: Digest,
-    hash_chain_type: HashChainType,
-) -> Result<(), BridgeConstraintsError> {
-    let rebuilt_hash_chain =
-        hashes.fold(prev_hash_chain, |acc, hash| keccak256_combine([acc, hash]));
-
-    if rebuilt_hash_chain != new_hash_chain {
-        return Err(BridgeConstraintsError::MismatchHashChain {
-            computed: rebuilt_hash_chain,
-            input: new_hash_chain,
-            hash_chain_type,
-        });
-    }
+    /// Validate that the rebuilt hash chain is equal to the new hash chain.
+    fn validate_hash_chain(
+        &self,
+        hashes: &[Digest],
+        prev_hash_chain: Digest,
+        new_hash_chain: Digest,
+        hash_chain_type: HashChainType,
+    ) -> Result<(), BridgeConstraintsError> {
+        let rebuilt_hash_chain = hashes
+            .iter()
+            .fold(prev_hash_chain, |acc, &hash| keccak256_combine([acc, hash]));
+
+        if rebuilt_hash_chain != new_hash_chain {
+            eprintln!(
+                "block_hash. prev: {:?}, new: {:?}",
+                self.prev_l2_block_hash, self.new_l2_block_hash
+            );
+            hashes.iter().enumerate().for_each(|(idx, hash)| {
+                eprintln!("element #{idx} ({:?}): {hash:?}", hash_chain_type)
+            });
```

Comment:
Nit: you should probably combine this to one log line. This way it would not be easily readable..

Reply:
those print outs are directly in stderr by the sp1 zkvm, outside of tracing and so are printed in a readable consecutive way:

```
      [aggkit-prover] stderr: element #0 (InsertedGER): 29376da9927d334f42dcfab72e66eaa680b27885ee051e641cf873a624f519a7
      [aggkit-prover] stderr: element #1 (InsertedGER): 09ee5a1fa8eba4cbb5c33073a977fa25dea49f253e24f2c0579de5127de6bd70
      [aggkit-prover] stderr:
      [aggkit-prover] stderr: thread '<unnamed>' panicked at src/main.rs:9:82:
      [aggkit-prover] stderr: called `Result::unwrap()` on an `Err` value: BridgeConstraintsError(MismatchHashChain { 
```

Reply:
Ok. I still have mixed feelings about it, but I'll approve.